File: music-api/src/main/java/com/music/sale/common/ApiResponse.java
package com.music.sale.common;

import com.fasterxml.jackson.annotation.JsonInclude;

@JsonInclude(JsonInclude.Include.NON_NULL)
public class ApiResponse<T> {

    private final boolean success;
    private final T data;
    private final String message;
    private final String code;

    private ApiResponse(boolean success, T data, String message, String code) {
        this.success = success;
        this.data = data;
        this.message = message;
        this.code = code;
    }

    public static <T> ApiResponse<T> success(T data) {
        return success(data, null);
    }

    public static <T> ApiResponse<T> success(T data, String code) {
        return new ApiResponse<>(true, data, null, code);
    }

    public static <T> ApiResponse<T> error(String message) {
        return error(message, null, null);
    }

    public static <T> ApiResponse<T> error(String message, String code) {
        return error(message, code, null);
    }

    public static <T> ApiResponse<T> error(String message, String code, T data) {
        return new ApiResponse<>(false, data, message, code);
    }

    public boolean isSuccess() {
        return success;
    }

    public T getData() {
        return data;
    }

    public String getMessage() {
        return message;
    }

    public String getCode() {
        return code;
    }
}

File: src/main/kotlin/com/music.sale/adapter/web/common/ApiResponse.kt
package com.music.sale.adapter.web.common

data class ApiResponse<T>(
    val success: Boolean,
    val code: String,
    val message: String? = null,
    val data: T? = null
) {
    companion object {
        fun <T> success(data: T, code: String = "SUCCESS") = ApiResponse(
            success = true,
            code = code,
            data = data
        )

        fun <T> error(code: String, message: String? = null) = ApiResponse<T>(
            success = false,
            code = code,
            message = message
        )
    }
} 